substratevm/src/com.oracle.svm.hosted/src/com/oracle/svm/hosted/image/NativeImageDebugInfoProvider.java

```diff
@@ -1061,10 +1062,16 @@ public Path cachePath() {
 
         @Override
         public ResolvedJavaType ownerType() {
+            ResolvedJavaType result;
             if (method instanceof HostedMethod) {
-                return getDeclaringClass((HostedMethod) method, true);
+                result = getDeclaringClass((HostedMethod) method, true);
```

Comment:
`getDeclaringClass((HostedMethod) method, true)` calls getOriginal(hostedMethod.getDeclaringClass()) which contains
```java
private static ResolvedJavaType getOriginal(HostedType hostedType) {                 
    /* partially unwrap then traverse through substitutions to the original */       
    ResolvedJavaType javaType = hostedType.getWrapped().getWrappedWithoutResolve();  
    if (javaType instanceof SubstitutionType) {                                      
        return ((SubstitutionType) javaType).getOriginal();                          
    } else if (javaType instanceof CustomSubstitutionType<?, ?>) {                   
        return ((CustomSubstitutionType<?, ?>) javaType).getOriginal();              
    } else if (javaType instanceof LambdaSubstitutionType) {                         
        return ((LambdaSubstitutionType) javaType).getOriginal();                    
    } else if (javaType instanceof InjectedFieldsType) {                             
        return ((InjectedFieldsType) javaType).getOriginal();                        
    }                                                                                
    return javaType;                                                                 
}                                                                                    
```
Instead of calling hostedType.getWrapped().**getWrappedWithoutResolve()** and manually figuring out what substitution this could be, wouldn't it be equivalent to just call hostedType.getWrapped().**getWrapped()**. This way, getting to the originals is performed via the SubstitutionProcessors responsible for the substitutions in the first place. Or am I missing something? cc @adinn

Reply:
Thinking about it, wouldn't the following simpler implementation also do the job:
```java
public ResolvedJavaType ownerType() {
    ResolvedJavaType result = method.getDeclaringClass();
    while (result instanceof WrappedJavaType) {
        result = ((WrappedJavaType) result).getWrapped();
    }
    return result;
}

Reply:
@olpaw Yes, the partial unwrapping is needed. In the case where we the wrapping is

    HostedType -> AnalysisType -> SubstitutionType

calling getWrapped().getWrapped() will result in a call to [AnnotationSubstitutionProcessor.resolve](https://github.com/oracle/graal/blob/85736d89e39a6a4fb700c45c1bd092205ad4ae3f/substratevm/src/com.oracle.svm.hosted/src/com/oracle/svm/hosted/substitute/AnnotationSubstitutionProcessor.java#L131) and return the annotated type.

Reply:
> Thinking about it, wouldn't the following simpler implementation also do the job:
>
>     public ResolvedJavaType ownerType() {
>        ResolvedJavaType result = method.getDeclaringClass();
>        while (result instanceof WrappedJavaType) {
>            result = ((WrappedJavaType) result).getWrapped();
>        }
>        return result;
>     }

No, as noted above, that will return the annotated type when method.getDeclaringClass() is a SubstitutionType.

Reply:
I see. Thanks for clarification.